sim/druid/talents.go

```diff
@@ -90,6 +91,43 @@ func (druid *Druid) ApplyTalents() {
 	druid.applyPrimalMadness()
 }
 
+func (druid *Druid) applyNaturesGrace() {
+	if druid.Talents.NaturesGrace == 0 {
+		return
+	}
+
+	ngAuraSpellId := []int32{0, 16880, 61345, 61345}[druid.Talents.NaturesGrace]
+	ngAuraSpellHastePct := []float64{0, 0.05, 0.1, 0.15}[druid.Talents.NaturesGrace]
+	ngCooldown := core.Cooldown{
+		Timer:    druid.NewTimer(),
+		Duration: time.Second * 60,
+	}
+
+	ngAura := druid.RegisterAura(core.Aura{
+		Label:    "Natures Grace Proc",
+		ActionID: core.ActionID{SpellID: ngAuraSpellId},
+		Duration: time.Second * 15,
+		Icd:      &ngCooldown,
+		OnGain: func(aura *core.Aura, sim *core.Simulation) {
+			druid.MultiplyCastSpeed(1 + ngAuraSpellHastePct)
+		},
+		OnExpire: func(aura *core.Aura, sim *core.Simulation) {
+			druid.MultiplyCastSpeed(1 / (1 + ngAuraSpellHastePct))
+		},
+	})
+
+	core.MakeProcTriggerAura(&druid.Unit, core.ProcTrigger{
+		Name:           "Natures Grace (Talent)",
+		Callback:       core.CallbackOnApplyEffects,
```

Comment:
It may seem unintuitive but this has to be `CallbackOnSpellHitDealt`. We don't have `OnApplyOrFresh` callback and the `OnApplyEffects` trigger evaluator will only get a dummy outcome that will never match your target outcome, as the spells actually decide within this callback if the spell lands or not.

This is the reason you 'DealDamage' even for spells that have 0 initial damage. 
https://github.com/wowsims/cata/pull/245/files#diff-bb45471b50c62e10efbcb2d29519e67b22944940bbcf3c46fd77c8336a815a24R60

**Note**: This only the `OnSpellHitDealt` callback, so it won't proc off periodic damage ticks.

Reply:
resolved, thanks